### PR TITLE
test: improve e2e handling cypress screenshots to percy and e2e visual review check

### PR DESCRIPTION
**Pull request by johnhwhite**
[AB#2555499](https://dev.azure.com/blackbaud/f565481a-7bc9-4083-95d5-4f953da6d499/_workitems/edit/2555499)

### CODE REVIEW

**Comment by blackbaud-sky-build-user** [comment]
[Storybook preview](https://blackbaud.github.io/skyux-pr-preview/1392/storybook/)

Component Storybooks:

- [a11y-storybook](https://blackbaud.github.io/skyux-pr-preview/1392/storybooks/a11y-storybook/)
- [ag-grid-storybook](https://blackbaud.github.io/skyux-pr-preview/1392/storybooks/ag-grid-storybook/)
- [core-storybook](https://blackbaud.github.io/skyux-pr-preview/1392/storybooks/core-storybook/)
- [forms-storybook](https://blackbaud.github.io/skyux-pr-preview/1392/storybooks/forms-storybook/)
- [indicators-storybook](https://blackbaud.github.io/skyux-pr-preview/1392/storybooks/indicators-storybook/)
- [layout-storybook](https://blackbaud.github.io/skyux-pr-preview/1392/storybooks/layout-storybook/)
- [lists-storybook](https://blackbaud.github.io/skyux-pr-preview/1392/storybooks/lists-storybook/)
- [modals-storybook](https://blackbaud.github.io/skyux-pr-preview/1392/storybooks/modals-storybook/)
- [pages-storybook](https://blackbaud.github.io/skyux-pr-preview/1392/storybooks/pages-storybook/)
- [popovers-storybook](https://blackbaud.github.io/skyux-pr-preview/1392/storybooks/popovers-storybook/)
- [split-view-storybook](https://blackbaud.github.io/skyux-pr-preview/1392/storybooks/split-view-storybook/)
- [storybook](https://blackbaud.github.io/skyux-pr-preview/1392/storybooks/storybook/)
- [tabs-storybook](https://blackbaud.github.io/skyux-pr-preview/1392/storybooks/tabs-storybook/)
- [theme-storybook](https://blackbaud.github.io/skyux-pr-preview/1392/storybooks/theme-storybook/)
- [tiles-storybook](https://blackbaud.github.io/skyux-pr-preview/1392/storybooks/tiles-storybook/)

Apps:

- [code-examples](https://blackbaud.github.io/skyux-pr-preview/1392/code-examples/)
- [dep-graph](https://blackbaud.github.io/skyux-pr-preview/1392/dep-graph/)
- [integration](https://blackbaud.github.io/skyux-pr-preview/1392/integration/)
- [playground](https://blackbaud.github.io/skyux-pr-preview/1392/playground/)

File: libs/sdk/e2e-schematics/src/lib/send-cypress-screenshots-to-percy/send-cypress-screenshots-to-percy.spec.ts [hunk 1]
```diff
@@ -0,0 +1,177 @@
+import { stripIndent } from '@angular-devkit/core/src/utils/literals';
+// eslint-disable-next-line @typescript-eslint/ban-ts-comment
+// @ts-ignore
+import utils from '@percy/sdk-utils';
```
Comment (johnhwhite) [inline libs/sdk/e2e-schematics/src/lib/send-cypress-screenshots-to-percy/send-cypress-screenshots-to-percy.spec.ts]: There are no types available for this library.

File: .github/workflows/e2e.yml [hunk 2]
```diff
@@ -247,14 +247,25 @@ jobs:
       - name: Percy ${{ matrix.project }}
         uses: actions/github-script@v6
         with:
-          retries: 1
```
Comment (johnhwhite) [inline .github/workflows/e2e.yml:250]: I misunderstood `retries` originally — it only applies to rest calls.

File: libs/sdk/e2e-schematics/src/lib/send-cypress-screenshots-to-percy/send-cypress-screenshots-to-percy.ts [hunk 3]
```diff
@@ -0,0 +1,108 @@
+import { join, normalize } from '@angular-devkit/core';
+import { stripIndent } from '@angular-devkit/core/src/utils/literals';
+// eslint-disable-next-line @typescript-eslint/ban-ts-comment
+// @ts-ignore
+import utils from '@percy/sdk-utils';
+
+import { readFileSync } from 'fs';
+
+export function sendCypressScreenshotsToPercy(
```
Comment (johnhwhite) [inline libs/sdk/e2e-schematics/src/lib/send-cypress-screenshots-to-percy/send-cypress-screenshots-to-percy.ts:7]: This function consolidates a pattern we used in a couple libraries and adds test coverage.

File: libs/sdk/e2e-schematics/src/lib/send-cypress-screenshots-to-percy/send-cypress-screenshots-to-percy.spec.ts (hunk 1 above)
Comment (Blackbaud-SteveBrush) [reply libs/sdk/e2e-schematics/src/lib/send-cypress-screenshots-to-percy/send-cypress-screenshots-to-percy.spec.ts]: Paul started a convention to support packages without types. See the `/vendor/types/*` directory.

File: libs/sdk/e2e-schematics/src/lib/send-cypress-screenshots-to-percy/send-cypress-screenshots-to-percy.ts [hunk 4]
```diff
@@ -0,0 +1,108 @@
+import { join, normalize } from '@angular-devkit/core';
+import { stripIndent } from '@angular-devkit/core/src/utils/literals';
+// eslint-disable-next-line @typescript-eslint/ban-ts-comment
+// @ts-ignore
+import utils from '@percy/sdk-utils';
```
Comment (Blackbaud-SteveBrush) [inline libs/sdk/e2e-schematics/src/lib/send-cypress-screenshots-to-percy/send-cypress-screenshots-to-percy.ts:3]: Ditto.

File: package.json [hunk 5]
```diff
@@ -145,6 +145,7 @@
     "@percy/cli": "1.24.0",
     "@percy/core": "1.24.0",
     "@percy/cypress": "3.1.2",
+    "@percy/sdk-utils": "^1.24.2",
```
Comment (Blackbaud-SteveBrush) [inline package.json]: Caret.

File: vendor/types/@percy/index.d.ts [hunk 6]
```diff
@@ -0,0 +1 @@
+declare module '@percy';
```
Comment (Blackbaud-SteveBrush) [inline vendor/types/@percy/index.d.ts]: Is this file necessary?

Comment (johnhwhite) [reply vendor/types/@percy/index.d.ts]: Yes. Without it there were `error TS2688: Cannot find type definition file for '@percy'. The file is in the program because: Entry point for implicit type library '@percy'` errors during e.g. `nx build playground`.

Comment (Blackbaud-SteveBrush) [reply vendor/types/@percy/index.d.ts]: I know why. Delete `vendor/types/@percy/index.d.ts` file, and then modify `tsconfig.base.json` (it was only looking for declaration files one level deep):
```
"typeRoots": ["./node_modules/@types", "./vendor/types/**/*.d.ts"],
```
Now the file in `vendor/types/@percy/sdk-utils/index.d.ts` should work.

**Review by Blackbaud-SteveBrush — DISMISSED** [review]

**Review by Blackbaud-TrevorBurch — APPROVED** [review]
